### PR TITLE
Image test improvements

### PR DESCRIPTION
**Pull request by ahuang98**
More commit history and comments from https://github.com/apache/kafka/pull/14206 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mimaison** [comment]
@ahuang98 The [last CI](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15373/3/#showFailuresLink) run for this PR had a lot of test failures. I think if you rebase on trunk that should clear most of them.

File: metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java [hunk 1]
```diff
@@ -186,6 +264,19 @@ public void testImage2RoundTrip() {
         testToImage(IMAGE2);
     }
 
+    public void testApplyDelta2() {
```
Comment (mimaison) [inline metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java:268]: Is this missing the `@Test` annotation?

File: metadata/src/test/java/org/apache/kafka/image/ImageDowngradeTest.java [hunk 2]
```diff
@@ -128,6 +128,39 @@ public void testPreControlledShutdownStateVersion() {
                         TEST_RECORDS.get(1)));
     }
 
+    /**
+     * Test downgrading to a MetadataVersion that doesn't support ZK migration.
+     */
+    @Test
+    public void testPreZkMigrationSupportVersion() throws Throwable {
```
Comment (mimaison) [inline metadata/src/test/java/org/apache/kafka/image/ImageDowngradeTest.java]: This does not throw so we can remove `throws Throwable`

**Review by mimaison — COMMENTED** [review]
Thanks for the PR. It looks good overall. I just left a couple of comments.

File: metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java (hunk 1 above)
Comment (mimaison) [reply metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java:268]: Sorry I should have added that after enabling that test, it is failing.
Expected:
`brokers([...], 1(BrokerRegistration(id=1, epoch=1001, incarnationId=U52uRe20RsGI0RvpcTx33Q, listeners=[Endpoint(listenerName='PLAINTEXT', securityProtocol=PLAINTEXT, host='localhost', port=9093)], supportedFeatures={foo: 1-3}, rack=Optional.empty, fenced=true, inControlledShutdown=true, isMigratingZkBroker=false, directories=[])), [...]`
Actual:
`brokers([...], 1(BrokerRegistration(id=1, epoch=1001, incarnationId=U52uRe20RsGI0RvpcTx33Q, listeners=[Endpoint(listenerName='PLAINTEXT', securityProtocol=PLAINTEXT, host='localhost', port=9093)], supportedFeatures={foo: 1-3}, rack=Optional.empty, fenced=false, inControlledShutdown=false, isMigratingZkBroker=false, directories=[])), [...]`
Broker 1 has `fenced=false` and `inControlledShutdown=false` while we expected both to be `true`.

**Review by mimaison — APPROVED** [review]
LGTM

File: metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java [hunk 3]
```diff
@@ -88,7 +101,7 @@ public class ClusterImageTest {
             setId(2).
             setEpoch(123).
             setIncarnationId(Uuid.fromString("hr4TVh3YQiu3p16Awkka6w")).
-            setListeners(Arrays.asList(new Endpoint("PLAINTEXT", SecurityProtocol.PLAINTEXT, "localhost", 9093))).
+            setListeners(Arrays.asList(new Endpoint("PLAINTEXT", SecurityProtocol.PLAINTEXT, "localhost", 9094))).
```
Comment (chia7712) [inline metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java:104]: `ReplicaManagerTest` reuse the `IMAGE1` to test the listeners, hence this change break some test cases. Do you have free cycle to fix it? If not, I can file PR to fix it :)

Comment (chia7712) [reply metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java:104]: I have filed a ticket https://issues.apache.org/jira/browse/KAFKA-16447

please feel free to take over it

Comment (jolshan) [reply metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java:104]: I just found this as well. I opened https://issues.apache.org/jira/browse/KAFKA-16451 but will close as a duplicate

Comment (ahuang98) [reply metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java:104]: Thanks folks, will open up a fix soon

Comment (ahuang98) [reply metadata/src/test/java/org/apache/kafka/image/ClusterImageTest.java:104]: Responded on https://issues.apache.org/jira/browse/KAFKA-16447, someone is volunteering to fix, will check in again tomorrow!

**Pull request merged**